### PR TITLE
plugin, config: make license key entirely optional

### PR DESCRIPTION
**Pull request by dgw**
Tin. Built on #4, but takes it to a new extreme as discussed in https://github.com/sopel-irc/sopel-iplookup/issues/3#issuecomment-1989629080.

Motivated by some joker on AWS repeatedly exhausting our account's daily download quota, spamming me with email notifications from MM, plus breaking any new installs of Sopel 7.x and this standalone plugin for 8.x.

Since the daily quota appears to be around 32 downloads, the existing download solution with an inbuilt key was never going to scale particularly well anyway. We might as well prepare now.

### CODE REVIEW

**Comment by dgw** [comment]
Note: Starting off in draft state because I can't actually end-to-end test the download-with-key code path while our quota's exhausted. If abuse of the published key continues and I still can't test after a day or two, I'll see about creating a separate personal account at MM just for this. (Hopefully not necessary.)

**Comment by dgw** [comment]
I successfully tested the direct-download code path at 1 minute past midnight UTC. Noticed a small opportunity for logging improvements, so did that too.

**Comment by dgw** [comment]
Rebased after #4 merged.

File: README.rst
Comment (SnoopJ) [inline README.rst:31]: I think that's meant to be a code block?

````suggestion
```
$ sopel-plugins configure iplookup
Configure Sopel GeoIP Lookup Plugin
Please consult sopel-iplookup's README to learn about its settings.

Path to existing GeoIP db files (leave empty to auto download): 
MaxMind license key (optional): (hidden input)
```
````

File: README.rst
Comment (SnoopJ) [inline README.rst:40]: Hmm, the wording here bugs me a little bit. I think I'd be happier if the information here was in the reverse order. I.e. tell the user that the plugin has a sane default for GeoIP data, _then_ mention the alternate ways to supply a particular database.

File: sopel_iplookup/plugin.py
Comment (SnoopJ) [inline sopel_iplookup/plugin.py:128]: This looks like we should have a separate `_download_geoip()` helper (or two) to contain this new code

**Review by SnoopJ — CHANGES_REQUESTED** [review]
Changes look generally good, I did have a few remarks.

File: README.rst
Comment (dgw) [reply README.rst:31]: This is rST, not Markdown.

I'm not against changing the readme to Markdown, though. It'd be simpler to test-render with `grip` or a VS Code extension. (Couldn't get the one I found for rST to work.)

File: README.rst
Comment (dgw) [reply README.rst:40]: Really, at this point I'd be surprised if I managed to write something _without_ second-guessing the information order. Seems to be a tradition for me writing docs that you later review 😂

File: sopel_iplookup/plugin.py
Comment (dgw) [reply sopel_iplookup/plugin.py:128]: Maybe. I chose not to be afraid of the extra indentation level, since much of the logic is shared. How about a `_get_geoip_urls()` helper?

File: README.rst
Comment (SnoopJ) [reply README.rst:31]: Sigh, right, my classic mistake 🤦

It does render incorrectly, though, I guess it needs an extra colon in the preceding paragraph

image

File: sopel_iplookup/plugin.py
Comment (SnoopJ) [reply sopel_iplookup/plugin.py:128]: I tend to drop leading `get` from helpers like that but I don't feel strongly about it. I don't mind the extra indentation level, but collapsing 24 lines to _"get the relevant download URL"_ feels like a win for legibility to me here.

File: README.rst
Comment (dgw) [reply README.rst:31]: Oh, so it does. Without the double colon it's taken as a blockquote, I guess.

File: README.rst
Comment (dgw) [reply README.rst:40]: This is going to come with a rewrite, since just swapping the paragraphs didn't flow very well.

File: README.rst
Comment (dgw) [reply README.rst:31]: Should be fixed in the last push.

File: sopel_iplookup/plugin.py
Comment (dgw) [reply sopel_iplookup/plugin.py]: Functions and methods should always have verb-y names, in my book, but if it helps: I didn't use `get`. 😁

File: README.rst
Comment (SnoopJ) [inline README.rst:40]: Nice, I like this rewording. Thank you for indulging me 😁

**Review by SnoopJ — APPROVED** [review]
Nice one, let's do it :shipit:

**Comment by dgw** [comment]
> Nice one, let's do it :shipit:

I'll shortly see about a PR w/automated release workflow triggered by `published` Releases on GH. Need anything else from me?

**Comment by SnoopJ** [comment]
Nope, nothing from me

**Comment by dgw** [comment]
As promised on IRC, I checked if this needed any squashing. I eliminated a fixup commit, but the rest of the history looked sane enough. Should be good to merge when you are back in decent-connectivity-land.

**Comment by dgw** [comment]
@SnoopJ _Daily database download limit reached_ again. I'd like to delete the existing API key soon, but first we need a new release of this plugin available that doesn't need it. This patch does that.

I've been holding off because this plugin's nominally your baby, but I'm happy to merge and finish up a new release if you aren't able to spend time on it. 😸

**Pull request merged**
